src/System.Private.CoreLib/src/System/StubHelpers.cs

```diff
@@ -1715,14 +1720,28 @@ internal static void CheckStringLength(int length)
 
         internal static void CheckStringLength(uint length)
         {
-            if (length > 0x7ffffff0)
+            if (length > MaxStringLength)
             {
-                throw new MarshalDirectiveException(SR.Marshaler_StringTooLong);
+                ThrowStringTooLong();
             }
         }
 
-        [MethodImplAttribute(MethodImplOptions.InternalCall)]
-        internal static extern unsafe int strlen(sbyte* ptr);
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static unsafe int strlen(byte* ptr)
+        {
+            int length = SpanHelpers.IndexOf(ref *ptr, (byte)'\0', MaxStringLength);
```

Comment:
This can use comment that this depends on `IndexOf` processing memory in aligned chunks, and thus it won't crash even if it access memory beyond the null terminator.